src/realizations/catchment/Bmi_Py_Formulation.cpp

```diff
@@ -7,10 +7,154 @@ using namespace models::bmi;
 
 using namespace pybind11::literals;
 
+Bmi_Py_Formulation::Bmi_Py_Formulation(std::string id, Forcing forcing, utils::StreamHandler output_stream)
+: Bmi_Module_Formulation<models::bmi::Bmi_Py_Adapter>(id, forcing, output_stream) { }
+
+Bmi_Py_Formulation::Bmi_Py_Formulation(std::string id, forcing_params forcing_config, utils::StreamHandler output_stream)
+: Bmi_Module_Formulation<models::bmi::Bmi_Py_Adapter>(id, forcing_config, output_stream) { }
+
+shared_ptr<Bmi_Py_Adapter> Bmi_Py_Formulation::construct_model(const geojson::PropertyMap &properties) {
+    auto python_type_name_iter = properties.find(BMI_REALIZATION_CFG_PARAM_OPT__PYTHON_TYPE_NAME);
+    if (python_type_name_iter == properties.end()) {
+        throw std::runtime_error("BMI Python formulation requires Python model class type, but none given in config");
+    }
+    std::string python_type_name = python_type_name_iter->second.as_string();
+
+    return std::make_shared<Bmi_Py_Adapter>(
+            Bmi_Py_Adapter(
+                    get_model_type_name(),
+                    get_bmi_init_config(),
+                    python_type_name,
+                    get_allow_model_exceed_end_time(),
+                    is_bmi_model_time_step_fixed(),
+                    output));
+}
+
+time_t realization::Bmi_Py_Formulation::convert_model_time(const double &model_time) {
+    return (time_t) (get_bmi_model()->convert_model_time_to_seconds(model_time));
+}
+
+const vector<string> Bmi_Py_Formulation::get_bmi_input_variables() {
+    return get_bmi_model()->GetInputVarNames();
+}
+
+const vector<string> Bmi_Py_Formulation::get_bmi_output_variables() {
+    return get_bmi_model()->GetOutputVarNames();
+}
+
 std::string Bmi_Py_Formulation::get_formulation_type() {
     return "bmi_py";
 }
 
+string Bmi_Py_Formulation::get_output_line_for_timestep(int timestep, std::string delimiter) {
+    // TODO: something must be added to store values if more than the current time step is wanted
+    // TODO: if such a thing is added, it should probably be configurable to turn it off
+    if (timestep != (next_time_step_index - 1)) {
+        throw std::invalid_argument("Only current time step valid when getting output for BMI Python formulation");
+    }
+    std::string output_str;
+
+    for (const std::string& name : get_output_variable_names()) {
+        output_str += (output_str.empty() ? "" : ",") + std::to_string(get_var_value_as_double(name));
+    }
+    return output_str;
+}
+
+double Bmi_Py_Formulation::get_response(time_step_t t_index, time_step_t t_delta) {
+    if (get_bmi_model() == nullptr) {
+        throw std::runtime_error("Trying to process response of improperly created BMI C formulation.");
```

Comment:
This error message is not correct, this is a Python formulation, not a C formulation.